go.mod

```diff
@@ -1,6 +1,6 @@
 module github.com/containerd/containerd/v2
 
-go 1.20
+go 1.21
```

Comment:
Are we good dropping support for go1.20 (which is still a supported version of go)? This will refuse the module to be used with older go versions. I'm honestly still somewhat uncomfortable doing so.

Reply:
go 1.20 apparently goes out of support by end Jan 2024 and upstream k8s team is not willing to consider using 1.20 . If we want to use v1.29, I am not sure there is another alternative though :( 

Thoughts? @dmcgowan @mikebrow @thaJeztah

Reply:
I guess the question becomes if we were to check this in today we'd be dropping support for 1.20 about a month early, have we done this prior?

Reply:
Looks like we moved off 1.19 in August, and EOL was September so there's one at least https://github.com/containerd/containerd/pull/8957

Reply:
I think the bigger question is not about us as a project as I think we move reasonably quickly onto new Go releases. What it sounds like (correct me if I'm wrong) is that merging this change means no downstream builder of containerd (e.g. distros) could use anything less than Go 1.21, which might be a reach for distros at this time? I don't know if that is @thaJeztah's concern, but I guess this is an overall industry question, as I assume K8s 1.29 is going to force the issue across a larger set of projects than just containerd?

Reply:
yeah moving to 1.29 would cause a wider issue with many projects.

@dmcgowan k8s 1.29 brings support to pass runtimeHandler through CRI. Do you have any thoughts on how we want to upgrade to 1.29?

Reply:
go 1.22 is "expected" in Feb... not guaranteed..

by https://github.com/kubernetes/release/issues/3347 ... I take it k8s decided to build cross on 1.20 and 1.21 but then also changed the go mod dependency of the apis to 1.21 in the k8s v1.29 release back in nov

I saw a cve.. but that is in service on the later 1.20.x so..

Reply:
Is there a way we want to move forward with updating to 1.29? It is blocking work on KEP 4216 (Image pull per runtime class). 1.29 has the changes to pass the runtime handler via cri to containerd.

Cc @dims - for status of kep 4216

Reply:
hmm  appears to be confirmed that kubernetes has moved to latest in the past here go 1.19
https://github.com/containerd/containerd/blob/release/1.7/go.mod
here go1.20
https://github.com/containerd/containerd/blob/release/1.7/vendor/modules.txt#L793C1-L794C21

so not new? 

Kirtana could you check if you can keep to our go mod as it was on 1.20 and redo the make vendor && verify-vendor

Reply:
you might have to do a clean build after?

Reply:
@mikebrow if you are referring to go clean cache / mod cache I tried doing that when submitting the PR too. doing a go mod tidy and vendor does update the golang version in go.mod. Do you know how this can be avoided?

Reply:
https://github.com/containerd/containerd/pull/9378/files#diff-452178a65ac0c170166dc52fe02057fd71eb9abb3948616847f6a0b64544fb60R1-R2  

wondering if that is the problematic change?

Reply:
rebased and updated the branch after a clean cache as well. There is a hard requirement for 1.21 that I do not see a workaround for

Reply:
ok.. klog . slog https://github.com/go-logr/logr/pull/205

can we drop klog

Reply:
checking... looks like integration/ (code pretending to be like kubelet) and some cri exec/streaming code is tied to it.. based on some legacy cut-n-paste code

Reply:
from the quick read of the above ^ linked 205 pr it seems the desire to use a 1.21 coded feature drove the dependency build check..